### DISCUSSION TITLE
Blurry image when Resizing and Extending

### DISCUSSION THREAD

**Discussion by tomasr78** [dlemstra/Magick.NET#1034] [Help] [answered]
I get blurry images in PDF when making them smaller and fitting to the PDF page in the center. 

```
using (var imageCollection = new MagickImageCollection())
{
         imageCollection.Add(MagickImage(@"C:\myfile\jun_pdf_image.png"));
         imageCollection.Add(MagickImage(@"C:\myfile\original.jpg"));                                
         imageCollection.Write(@"C:\Projects\_temp\output.pdf");
}

 private static MagickImage MagickImage(string imageFile)
{
            var image = new MagickImage();
            var magickReadSettings = new MagickReadSettings();
            image.Read(imageFile, magickReadSettings);

            var fromPageSize = MagickGeometry.FromPageSize("a4");

            image.Resize(fromPageSize.Width, fromPageSize.Height);
            image.Extent(fromPageSize.Width, fromPageSize.Height, Gravity.Center);                                   

            image.Page = fromPageSize;

            return image;
}


```
If I remove the `Resize `and `Extend `methods then the images look sharp as original in PDF but then images are not centered in PDF. The Extend does not work without Resize and Resize produces a blurry image.  Spends several hours to understand what's is causing blur in resized(downsized) images without any luck.

I have attached demo images and outputs, the output.pdf has centered but blurry images, the output-no-resize-no-extend.pdf has perfect quality images as I need the to images are no longer centered on the PDF page.

[images.zip](https://github.com/dlemstra/Magick.NET/files/7241906/images.zip)
[output.pdf](https://github.com/dlemstra/Magick.NET/files/7241891/output.pdf)
[output-no-resize-no-extend.pdf](https://github.com/dlemstra/Magick.NET/files/7241892/output-no-resize-no-extend.pdf)

**Comment by dlemstra** [comment, marked as answer]
When you resize the image you will make it smaller. When you want the best quality image you should only `Extent` the image. You should calculate what the width and height should be to have the same aspect ratio as the page size. You can calculate that with the following method (there might be a better method to do this):

```C#
var fromPageSize = MagickGeometry.FromPageSize("a4");

var width = image.Width / (double)fromPageSize.Width;
var height = image.Height / (double)fromPageSize.Height;
var max = Math.Max(width, height);
var newWidth = (int)(fromPageSize.Width * max);
var newHeight = (int)(fromPageSize.Height * max);

image.Extent(newWidth, newHeight, Gravity.Center);
```

**Reply by sefakpsz** [reply]
If I don't use resize and extent method together I cannot fit my image into my canvas but when I do that it has blur. Here is my code that I applied:

               `using var bitmap = SKBitmap.FromImage(image);
                // using var image2 = new MagickImage(outputStream.ToArray());
                using var memoryStream = new MemoryStream();
                bitmap.Encode(memoryStream, SKEncodedImageFormat.Png, 100);
                memoryStream.Position = 0;

                using var magickImage = new MagickImage(memoryStream, new MagickReadSettings { Format = MagickFormat.Png, BackgroundColor = MagickColors.Transparent, AntiAlias = true });

                var width2 = magickImage.Width / canvasWidth;
                var height2 = magickImage.Height / canvasHeight;
                var max = Math.Max(width2, height2);
                var newWidth = canvasWidth * width2;
                var newHeight = canvasHeight * height2;

                magickImage.Extent((uint)newWidth, (uint)newHeight, Gravity.Center);
                magickImage.Resize((uint)width, (uint)height);

                using var resizedStream = new MemoryStream();
                magickImage.Write(resizedStream);
                resizedStream.Position = 0;

                using var resizedImage = SKBitmap.Decode(resizedStream);`
